### PR TITLE
Hugging Face Hub Integration

### PR DESCRIPTION
**Pull request by jacobmarks**
Integrates FiftyOne Datasets with the Hugging Face Hub, with two utility functions: `push_to_hub()` and `load_from_hub()`.

## `push_to_hub()`

This utility takes your local dataset and
- creates a HF repo
- uploads the data to the repo (with FiftyOne's `dataset.export()`
- autogenerates a `DatasetCard` for the repo

### Usage

```py
import fiftyone as fo
import fiftyone.zoo as foz
import fiftyone.utils.hf_hub as fouh

dataset = foz.load_zoo_dataset("quickstart")
view = dataset[:10]

# upload view to HF hub
fouh.push_to_hub(view, "quickstart-10", private=True)
```

## `load_from_hub()`

This utility allows you to download and load any FiftyOne dataset that has been uploaded to the hub with `push_to_hub()`, as well as other Hugging Face datasets!

In particular, any HF repo with ` fiftyone.py` file (even if it is empty!) can be loaded in this fashion. By default, the `fouh.DefaultHuggingFaceLoader` is used to load the dataset from the repo, but the dataset author can define any subclass of the `fouh.BaseHuggingFaceLoader` class, and this will be used to perform the loading/conversion. In particular, the loader just needs to implement the `load(self)` method.

### Load a FiftyOne dataset that was pushed to the hub

```py
import fiftyone.utils.hf_hub as fouh

# load from HF hub
qs10 = fouh.load_from_hub(
    "<your-hf-username>/quickstart-10", 
    name = "my-hf-dataset",
    persistent=True
)

# launch app
session = fo.launch_app(qs10)
```

### Load a Hugging Face dataset that is compatible with the default loader:
```py
import fiftyone.utils.hf_hub as fouh

dataset = fouh.load_from_hub("jamarks/pokemon_copy", name="pokemon")
```

### Load a Hugging Face dataset that has a custom loader implemented in the repo's `fiftyone.py` file:
```py
import fiftyone.utils.hf_hub as fouh

dataset = fouh.load_from_hub("jamarks/fashionpedia-copy")
```

### Convert a Hugging Face dataset directly with default loader (if dataset's repo doesn't have a `fiftyone.py` file

```py
import fiftyone.utils.hf_hub as fouh
from datasets import load_dataset

hf_dataset = load_dataset("cppe-5", split='train')
loader = fouh.DefaultHuggingFaceLoader("cppe-5", hf_dataset)

dataset = loader.load()
```

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [x] Documentation: FiftyOne documentation changes
-   [x] Other

## Summary by CodeRabbit

- **New Features**
	- Enhanced Hugging Face integration, including dataset management through the Hugging Face Hub.
	- Added support for various tasks with Transformers models: image classification, object detection, semantic segmentation, monocular depth estimation, and zero-shot classification.
	- New utilities for pushing to and loading datasets from the Hugging Face Hub, creating dataset cards, and custom dataset conversions.

- **Documentation**
	- Updated documentation with detailed instructions on setting up Transformers, performing inference tasks, managing datasets on the Hub, and customizing dataset cards.

### CODE REVIEW

File: fiftyone/utils/hf_hub.py [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+import fiftyone.types as fot
+import huggingface_hub as hfh
+
+datasets = fou.lazy_import(
+    "datasets", callback=lambda: fou.ensure_package("datasets")
+)
+
+DATASET_CONTENT_TEMPLATE = """
+
+This is a [FiftyOne](https://github.com/voxel51/fiftyone) dataset with {num_samples} samples.
```
Comment (davanstrien) [inline fiftyone/utils/hf_hub.py:32]: Really like the automatic dataset card generation!

File: fiftyone/utils/hf_hub.py [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+            self.hf_dataset, self.repo_id, **self.kwargs
+        )
+
+
+def _is_convertable_to_fiftyone(repo_id):
+    api = hfh.HfApi()
+    return api.file_exists(repo_id, "fiftyone.py", repo_type="dataset")
+
+
+def _load_fiftyone_dataset_from_hub(repo_id, **kwargs):
```
Comment (davanstrien) [inline fiftyone/utils/hf_hub.py:493]: Does this also work with parquet files? an increasing number of datasets are created using datasets and `push_to_hub` so will be stored as Parquet.

**Review by davanstrien — COMMENTED** [review]
Really nice to see progress on this, left a few comments/questions

File: fiftyone/utils/hf_hub.py [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+
+    dataset_summary = _generate_dataset_summary(repo_id, dataset)
+    if dataset_summary is not None:
+        card_inputs["dataset_summary"] = dataset_summary
+
+    card_data = hfh.DatasetCardData(**card_inputs)
+    return hfh.DatasetCard.from_template(card_data)
+
+
+def push_to_hub(dataset, repo_name, private=True, **data_card_kwargs):
```
Comment (brimoor) [inline fiftyone/utils/hf_hub.py:121]: just a reminder to add docstrings to `push_to_hub()` and `load_from_hub()` before merging

File: fiftyone/utils/hf_hub.py [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+    else:
+        raise ValueError(f"Unknown dtype {dtype}")
+
+
+def _is_classification_field(feature):
+    return isinstance(feature, datasets.features.ClassLabel)
+
+
+def _add_classification_field(fo_dataset, hf_dataset, feature_name):
+    print(f"Adding classification field {feature_name} to dataset")
```
Comment (brimoor) [inline fiftyone/utils/hf_hub.py:268]: reminder to remove `print()` statements before merging, unless you are intentionally including them, in which case our convention is to use the `logging` module as follows so users can configure whether/how logging is done:

``py
logger = logging.getLogger(__name__)

logger.info("your message")
```

File: fiftyone/utils/hf_hub.py [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+        conversion_module, BaseHuggingFaceLoader
+    )
+    if len(subclasses) == 0:
+        loader = DefaultHuggingFaceLoader(repo_id, hf_dataset, **kwargs)
+    else:
+        loader = subclasses[0](repo_id, hf_dataset, **kwargs)
+    return loader.load()
+
+
+def load_from_hub(repo_id, *args, **kwargs):
```
Comment (brimoor) [inline fiftyone/utils/hf_hub.py:523]: code readability suggestion: move `push_to_hub()` and `load_from_hub()` to the top of the module so anyone who wants to see what the module "does" will find the public API at the very top of the module.

Note that function declaration order does not matter in Python like it does in, say, C.

**Review by brimoor — COMMENTED** [review]
@jacobmarks just leaving a few cleanup suggestions; overall the proposed interface LGTM.

Please add documentation to `docs/integrations/huggingface.rst` before merging to make this integration "offical" 🤗

File: fiftyone/utils/hf_hub.py (hunk 4 above)
Comment (jacobmarks) [reply fiftyone/utils/hf_hub.py:268]: Got it. Didn't realize this, will do from now on

**Review by ehofesmann — COMMENTED** [review]
This is so cool! Made my first HuggingFace dataset :muscle: 

I had similar comments to Brian, the one request I'd like to add is an example `fiftyone.py` implementation in the docs page.

**Comment by jacobmarks** [comment]
> @jacobmarks just leaving a few cleanup suggestions; overall the proposed interface LGTM.
> 
> Please add documentation to `docs/integrations/huggingface.rst` before merging to make this integration "offical" 🤗

Thanks @brimoor ! Yeah I was waiting to add documentation for two reasons:
1. Wanted to make sure this interface was cool with you, and
2. The Depth Estimation PR is modifying the same file
Will add momentarily

File: fiftyone/utils/hf_hub.py [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+        export_dir=tmp_dir,
+        dataset_type=fot.FiftyOneDataset,
+        export_media=True,
+    )
+
+    hf_username = hfh.whoami()["name"]
+
+    ## Create the dataset repo
+    repo_id = os.path.join(hf_username, repo_name)
+    hfh.create_repo(repo_id, repo_type="dataset", private=private)
```
[the remark this answers is not in the document]
Comment (jacobmarks) [reply fiftyone/utils/hf_hub.py:135]: Adding in support for `exist_ok` argument.

**Comment by jacobmarks** [comment]
> This is so cool! Made my first HuggingFace dataset 💪
> 
> I had similar comments to Brian, the one request I'd like to add is an example `fiftyone.py` implementation in the docs page.

Thanks @ehofesmann ! Absolutely I can add that :)

**Comment by jacobmarks** [comment]
[hf_hub_integration_docs.pdf](https://github.com/voxel51/fiftyone/files/14254477/hf_hub_integration_docs.pdf)

@brimoor @ehofesmann I added some basic documentation for the HF hub integration. I'm sure there's a lot more we could do, but at least this is a starting point

**Comment by swheaton** [comment]
experimental ai reviewer ...

@coderabbitai review

File: fiftyone/utils/hf_hub.py [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+        repo_id=repo_id,
+        repo_type="dataset",
+    )
+
+    ## Create the dataset card
+    card = _create_dataset_card(repo_id, dataset, **data_card_kwargs)
+    card.push_to_hub(repo_id)
+
+    ## Clean up
+    shutil.rmtree(tmp_dir)
```
[the remark this answers is not in the document]
Comment (swheaton) [reply fiftyone/utils/hf_hub.py:87]: Good point ...
Even better ...

```python
with tempfile.TemporaryDirectory() as tmp_dir:
    ...
```

[the remark this answers is not in the document]
Comment (swheaton) [reply fiftyone/utils/hf_hub.py:87]: It responded to me!! Lol I feel special

**Comment by brimoor** [comment]
Closing in favor of https://github.com/voxel51/fiftyone/pull/4193

**Pull request closed without merging**